homeassistant/components/fan/template.py

```diff
@@ -0,0 +1,345 @@
+
+        self._state = False
+        self._speed = None
+        self._oscillating = None
+
+        self._template.hass = self.hass
+        if self._speed_template:
+            self._speed_template.hass = self.hass
+            self._supported_features |= SUPPORT_SET_SPEED
+        if self._oscillating_template:
+            self._oscillating_template.hass = self.hass
+            self._supported_features |= SUPPORT_OSCILLATE
+
+        self._entities = entity_ids
+        # List of valid speeds
+        self._speed_list = speed_list
+
+    @property
+    def name(self):
+        """Return the display name of this fan."""
+        return self._name
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def speed_list(self: ToggleEntity) -> list:
+        """Get the list of available speeds."""
+        return self._speed_list
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._state
+
+    @property
+    def speed(self):
+        """Return the current speed."""
+        return self._speed
+
+    @property
+    def oscillating(self):
+        """Return the oscillation state."""
+        return self._oscillating
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+    @asyncio.coroutine
+    def async_turn_on(self, speed: str=None) -> None:
+        """Turn on the fan.
+
+        This method is a coroutine.
+        """
+        self._state = True
+        yield from self.hass.async_add_job(self._on_script.async_run())
+
+        if speed:
+            yield from self.async_set_speed(speed)
+
+    @asyncio.coroutine
+    def async_turn_off(self) -> None:
+        """Turn off the fan.
+
+        This method is a coroutine.
+        """
+        self._state = False
+        yield from self.hass.async_add_job(self._off_script.async_run())
+
+    @asyncio.coroutine
+    def async_set_speed(self, speed: str) -> None:
+        """Set the speed of the fan.
+
+        This method is a coroutine.
+        """
+        if self._set_speed_script is None:
+            return
+
+        if speed in self._speed_list:
+            self._speed = speed
+            yield from self.hass.async_add_job(
+                self._set_speed_script.async_run({ATTR_SPEED: speed}))
+        else:
+            _LOGGER.error(
+                'Received invalid speed: %s. ' +
+                'Expected: %s.',
+                speed, self._speed_list)
+
+    @asyncio.coroutine
+    def async_oscillate(self, oscillating: bool) -> None:
+        """Set oscillation of the fan.
+
+        This method is a coroutine.
+        """
+        if self._set_oscillating_script is None:
+            return
+
+        if oscillating is True or oscillating is False:
+            self._oscillating = oscillating
+            yield from self.hass.async_add_job(
```

Comment:
`yield from self._*.async_run()`

Reply:
I'm interested in this component and would love to see it included. What's missing to make this PR go in?

@pvizeli any other code changes needed than moving statements like:

yield from self.hass.async_add_job(self._off_script.async_run())

to

yield from self._off_script.async_run()

? and testing it with latest home assistant master branch ...

Reply:
I reopened this PR here and removed usages of `async_add_job`. Please review
https://github.com/home-assistant/home-assistant/pull/12027
Thanks